src/main/java/com/redislabs/Connector.java

```diff
@@ -310,9 +316,106 @@ public void foreach(ArrayList<HashMap<String, Object>> record) throws Exception
         throw new Exception(msg, cause);
       }
     }
-    
   }
-  
+
+  private void retry(List<HashMap<String, Object>> record) throws Exception {
+    for(Map<String, Object> r: record) {
+      try {
+        Session session = connector.getSession();
+        Transaction transaction = session.beginTransaction();
+
+        Map<String, Object> value = (Map<String, Object>) r.get("value");
+        Map<String, Object> map = value.entrySet().stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue()));
+        String sourceName = (String) map.remove(SOURCE_STR);
+
+        String event = (String) map.remove(EVENT_STR);
+        if (event.charAt(0) != 'd') {
+          session.merge((String) map.remove(ENTETY_NAME_STR), map);
+        } else {
+          Source source = Source.getSource(sourceName);
+          Object o = session.get((String) map.remove(ENTETY_NAME_STR), (Serializable) map.get(source.getIdProperty().getName()));
+          // o can be null on hdel that removed the last field
+          if (o != null) {
+            session.delete(o);
+          }
+        }
+        transaction.commit();
+        session.clear();
+      } catch (Exception e) {
+        String msg = String.format("Failed retrying transaction error='%s'", e);
+        GearsBuilder.log(msg, LogLevel.WARNING);
+
+        //connector.closeSession();
+        if (connector.getSession().getTransaction().isActive()) {
+          connector.getSession().getTransaction().rollback();
+        }
+
+        connector.getSession().clear();
+        putInDLQ(r, e);
+      }
+    }
+  }
+
+  private void putInDLQ(Map<String, Object> r, Exception e) throws Exception {
+    String streamName = new String((byte[])r.get("key"));
+    Map<String, Object> map = (Map<String, Object>) r.get("value");
+
+    Map<String, String> newMap = new HashMap<>();
+
+    newMap.put("error", e.getMessage());
+    Throwable ex = e;
+    while (ex.getCause() != null)
+      ex = ex.getCause();
+    newMap.put("cause", ex.getLocalizedMessage());
+
+    String[] command;
+    Stream<String> commandStreamInit = Stream.of("XADD", DLQ_STR + streamName, "*");
```

Comment:
Does the stream name contains some `{}`, so we will know this stream will be in the same shard as the original one?

Reply:
It is the original stream name with a prefix of _DLQ_, so the hashtag is included.